plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs

```diff
@@ -688,20 +701,34 @@ processConstraint = \case
     MustPayToPubKeyAddress pk skhM mdv _refScript vl -> do
         -- TODO: implement adding reference script
         -- if datum is presented, add it to 'datumWitnesses'
+        networkId <- gets $ pNetworkId . cpsParams
         forM_ mdv $ \dv -> do
             unbalancedTx . tx . Tx.datumWitnesses . at (P.datumHash dv) .= Just dv
-        let hash = P.datumHash <$> mdv
-        unbalancedTx . tx . Tx.outputs %= (Tx.TxOut{ txOutAddress=pubKeyHashAddress pk skhM
-                                                   , txOutValue=vl
-                                                   , txOutDatumHash=hash
-                                                   } :)
+        let pv1TxOut = PV1.TxOut { PV1.txOutAddress=pubKeyHashAddress pk skhM
+                                 , PV1.txOutValue=vl
+                                 , PV1.txOutDatumHash=Nothing
+                                 }
+        let txInDatum = C.toCardanoTxOutDatumInTx mdv
+        let cardanoTxOut =
+              TxOut <$> C.toCardanoTxOut networkId C.toCardanoTxOutDatumHash pv1TxOut <&> outDatumHash .~ txInDatum
+        case cardanoTxOut of
+          Left err    -> throwError $ TxOutCardanoError err
```

Comment:
Should we have a function that gets the networkId, does the TxOut conversion and throws an error?